File: app/src/main/java/com/amazon/tv/leanbacklauncher/wallpaper/WallpaperImage.java
package com.amazon.tv.leanbacklauncher.wallpaper;

import android.content.Context;
import android.graphics.Matrix;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.widget.ImageView;

public class WallpaperImage extends ImageView {
    protected float mZoom;

    public WallpaperImage(Context context) {
        this(context, null);
    }

    public WallpaperImage(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public WallpaperImage(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        this.mZoom = 0.0f;
    }

    protected boolean setFrame(int l, int t, int r, int b) {
        boolean changed = super.setFrame(l, t, r, b);
        setScaleMatrix(getDrawable());
        return changed;
    }

    public void setZoomLevel(float zoom) {
        this.mZoom = zoom;
        setScaleMatrix(getDrawable());
        invalidate();
    }

    private void setScaleMatrix(Drawable drawable) {
        if (drawable != null) {
            float scale;
            float dx;
            Matrix matrix = getImageMatrix();
            int vwidth = (getWidth() - getPaddingLeft()) - getPaddingRight();
            int vheight = (getHeight() - getPaddingTop()) - getPaddingBottom();
            int dwidth = drawable.getIntrinsicWidth();
            int dheight = drawable.getIntrinsicHeight();
            if (dwidth * vheight > vwidth * dheight) {
                scale = (((float) vheight) / ((float) dheight)) * (this.mZoom + 1.0f);
                dx = (((float) vwidth) - (((float) dwidth) * scale)) * 0.5f;
            } else {
                scale = (((float) vwidth) / ((float) dwidth)) * (this.mZoom + 1.0f);
                dx = ((((float) vwidth) * this.mZoom) * 0.5f) * -1.0f;
            }
            float dy = ((((float) vheight) * this.mZoom) * 0.5f) * -1.0f;
            matrix.setScale(scale, scale);
            matrix.postTranslate((float) ((int) (dx + 0.5f)), (float) ((int) (dy + 0.5f)));
            setImageMatrix(matrix);
        }
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/wallpaper/WallpaperImage.kt
package com.amazon.tv.leanbacklauncher.wallpaper

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.widget.ImageView

open class WallpaperImage @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : androidx.appcompat.widget.AppCompatImageView(context, attrs, defStyle) {
    private var mZoom = 0.0f
    override fun setFrame(l: Int, t: Int, r: Int, b: Int): Boolean {
        val changed = super.setFrame(l, t, r, b)
        setScaleMatrix(drawable)
        return changed
    }

    fun setZoomLevel(zoom: Float) {
        mZoom = zoom
        setScaleMatrix(drawable)
        invalidate()
    }

    private fun setScaleMatrix(drawable: Drawable?) {
        if (drawable != null) {
            val scale: Float
            val dx: Float
            val matrix = imageMatrix
            val vwidth = width - paddingLeft - paddingRight
            val vheight = height - paddingTop - paddingBottom
            val dwidth = drawable.intrinsicWidth
            val dheight = drawable.intrinsicHeight
            if (dwidth * vheight > vwidth * dheight) {
                scale = vheight.toFloat() / dheight.toFloat() * (mZoom + 1.0f)
                dx = (vwidth.toFloat() - dwidth.toFloat() * scale) * 0.5f
            } else {
                scale = vwidth.toFloat() / dwidth.toFloat() * (mZoom + 1.0f)
                dx = vwidth.toFloat() * mZoom * 0.5f * -1.0f
            }
            val dy = vheight.toFloat() * mZoom * 0.5f * -1.0f
            matrix.setScale(scale, scale)
            matrix.postTranslate((dx + 0.5f).toInt().toFloat(), (dy + 0.5f).toInt().toFloat())
            imageMatrix = matrix
        }
    }
}